MSAL/test/integration/native_auth/end_to_end/otp_code_retriever/MSALNativeAuthEmailCodeRetriever.swift

```diff
@@ -0,0 +1,100 @@
+            return nil
+        }
+        let local = comps[0]
+        let domain = comps[1]
+
+        guard let lastMessageId = await retrieveLastMessage(local: local, domain: domain, retryCounter: numberOfRetry) else {
+            XCTFail("Something went wrong retrieving the messages for the email specified")
+            return nil
+        }
+        guard let emailOTPCode = await retrieveOTPCodeFromMessage(local: local, domain: domain, messageId: lastMessageId) else {
+            XCTFail("Something went wrong retrieving the OTP code")
+            return nil
+        }
+        return emailOTPCode
+    }
+
+    private func retrieveLastMessage(local: String, domain: String, retryCounter: Int) async -> Int? {
+        guard retryCounter > 0, let url = URL(string: baseURLString + "getMessages&login=\(local)&domain=\(domain)") else {
+            return nil
+        }
+        try? await Task.sleep(nanoseconds: UInt64(secondsToWait * Double(NSEC_PER_SEC)))
+        var request = URLRequest(url: url)
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        do {
+            let (data, response) = try await URLSession.shared.data(for: request)
+            
+            guard let httpResponse = response as? HTTPURLResponse,
+                    httpResponse.statusCode == 200,
+                    var dataDictionary = try JSONSerialization.jsonObject(with: data, options: []) as? [[String: Any]] else {
+                return nil
+            }
+            if dataDictionary.count > 0 {
+                dataDictionary.sort(by: {($0["date"] as? String ?? "") > ($1["date"] as? String ?? "")})
+                return dataDictionary.first?["id"] as? Int
+            } else {
+                // no emails found, retry
+                return await retrieveLastMessage(local: local, domain: domain, retryCounter: retryCounter - 1)
+            }
+        } catch {
+            print(error)
+            return nil
+        }
+    }
+
+    private func retrieveOTPCodeFromMessage(local: String, domain: String, messageId: Int) async -> String? {
+        guard let url = URL(string: baseURLString + "readMessage&login=\(local)&domain=\(domain)&id=\(messageId)") else {
+            return nil
+        }
+        var request = URLRequest(url: url)
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        do {
+            let (data, response) = try await URLSession.shared.data(for: request)
+
+            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
+                return nil
+            }
+            let dataDictionary = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
+            return (dataDictionary?["textBody"] as? String)?.components(separatedBy: CharacterSet.newlines).first(where: {$0.range(of: "[0-9]{4,}$", options: .regularExpression, range: nil, locale: nil) != nil })
```

Comment:
This is next to impossible to read or debug so if the position of the OTP changes it gets really complex to change this.
Maybe separate on multiple lines and add a comment?